pkg/kube/controllers/quarksstatefulset/active_passive_reconciler.go

```diff
@@ -0,0 +1,352 @@
+	for _, pod := range pods.Items {
+		if succeed, _ := r.execContainerCmd(&pod, container, cmd); succeed && podutil.IsPodReady(&pod) {
+			return &pod, nil
+		}
+	}
+	return nil, errors.Errorf("executing cmd: %v, in container: %s, failed\n", cmd, container)
+}
+
+func (r *ReconcileStatefulSetActivePassive) getActiveLabels(pods *corev1.PodList, container string) []corev1.Pod {
+	var labelledPods []corev1.Pod
+	for _, pod := range pods.Items {
+		labels := pod.GetLabels()
+		if _, found := labels[qstsv1a1.LabelActiveContainer]; found {
+			labelledPods = append(labelledPods, pod)
+		}
+	}
+	return labelledPods
+}
+
+func (r *ReconcileStatefulSetActivePassive) execContainerCmd(pod *corev1.Pod, container string, command []string) (bool, error) {
+	req := r.kclient.CoreV1().RESTClient().Post().
+		Resource("pods").
+		Name(pod.Name).
+		Namespace(pod.Namespace).
+		SubResource("exec").
+		VersionedParams(&corev1.PodExecOptions{
+			Container: container,
+			Command:   command,
+			Stdin:     true,
+			Stdout:    true,
+			Stderr:    false,
+		}, clientscheme.ParameterCodec)
+
+	executor, err := remotecommand.NewSPDYExecutor(r.kubeconfig, "POST", req.URL())
+	if err != nil {
+		return false, errors.New("failed to initialize remote command executor")
+	}
+	if err = executor.Stream(remotecommand.StreamOptions{Stdin: os.Stdin, Stdout: os.Stdout, Tty: false}); err != nil {
+		return false, errors.Wrapf(err, "failed executing command in pod: %s, container: %s in namespace: %s",
+			pod.Name,
+			container,
+			pod.Namespace,
+		)
+	}
+	ctxlog.Info(r.ctx, "Succesfully exec cmd in container: ", container, ", inside pod: ", pod.Name)
+
+	return true, nil
+}
+
+func (r *ReconcileStatefulSetActivePassive) getStsPodList(ctx context.Context, desiredSts *appsv1.StatefulSet) (*corev1.PodList, error) {
+	podList := &corev1.PodList{}
+	err := r.client.List(ctx,
+		podList,
+		crc.InNamespace(desiredSts.Namespace),
+	)
+	if err != nil {
+		return nil, err
+	}
+	return podList, nil
+}
+
+func (r *ReconcileStatefulSetActivePassive) listStatefulSetVersions(ctx context.Context, qStatefulSet *qstsv1a1.QuarksStatefulSet) (map[int]bool, error) {
+	ctxlog.Debug(ctx, "Listing StatefulSets owned by QuarksStatefulSet '", qStatefulSet.Name, "'.")
+
+	versions := map[int]bool{}
+
+	statefulSets, err := listStatefulSetsFromInformer(ctx, r.client, qStatefulSet)
+	if err != nil {
+		return nil, err
+	}
+
+	for _, statefulSet := range statefulSets {
+		strVersion, found := statefulSet.Annotations[qstsv1a1.AnnotationVersion]
+		if !found {
+			return versions, errors.Errorf("version annotation is not found from: %+v", statefulSet.Annotations)
+		}
+
+		version, err := strconv.Atoi(strVersion)
+		if err != nil {
+			return versions, errors.Wrapf(err, "version annotation is not an int: %s", strVersion)
+		}
+
+		ready, err := r.isStatefulSetReady(ctx, &statefulSet)
```

Comment:
shouldn't be required - there's 1 sts per qsts now - should just be able to retrieve it

Reply:
yes, that simplify my code